### ISSUE TITLE
Attatch to a windows service process using CodeXL

### ISSUE THREAD

**Issue by lonelygoat** [GPUOpen-Tools/CodeXL#40]
Hi there,

I need to attach to a already running process which is a windows service that uses opencl, and compiled using visual studio 2010. I am unable to find a way to attach to it using CodeXL so I can step into source code.

I failed to find related info in CodeXL getting started guide. Any hint that I can get the ball rolling ?

Thanks a lot....

**Comment by doronofek** [comment]
Short answer: The debugger cannot be attached to a running process.
Long answer:
This is because the debugger relies on DLL replacement to intercept the OpenCL API calls that the debugged process performs. If the debugger were attached to a running process that has already loaded the OpenCL runtime dll then the debugger would not be able to function.

**Comment by lonelygoat** [comment]
Thanks for your comments. However, I don't think debuggers rely on DLL replacement to intercept API calls generally speaking. there are many ways to intercept calls to a DLL(remember SoftICE, the most powerful debugger ever built). To attach to a running process is a standard feature for both Windows and Linux. Once loaded or attached to the executable, a debugger can change the code segment attribute and change the code any way it wants, such as writing a "int 3" instruction to the place where a break point is wanted.  Therefore, to load an executable or attach to a running executable is not a problem. The only problem for a programmer is to load symbols,  so one can debug at source level..Otherwise one can only read disassembly code with addresses and numbers.

CodeXL has a menu item called "attach", but it is always grayed somehow. I use vs2010 which has CodeXL plugin installed and corresponding CodeXL menu "attach" is grayed as well..

**Comment by lonelygoat** [comment]
Hi Doronofek

I looked at CodeXL manual and found that "attach to a process" is for profiling, not for debugging. Looks like you are correct in that CodeXL can not attach to a running process for debugging. If CodeXL is indeed implemented by replacing DLL to allow debugging, why it does not use so called "dll injection" instead ? although it is a bit ugly, all debuggers are using some ugly tricks anyway. Those tricks are not terribly difficult to implement. 

Now I built a testing environment by rewriting the windows service as a normal app, in which the testing app is loaded into CodeXL successfully...

Thanks a lot for shedding light on this...
